### PR TITLE
fix assertion on AuxPtr

### PR DESCRIPTION
**Pull request by leirocks**
While fixing a dead lock bug and adding assertion for that the work item queue should not be locked at the time of getting the auxPtrs. The assertion is wrong when JIT is collecting data in foreground thread causes stack probing and hence call dispose then re-enter script, which can call GetAuxPtrsWithLock. Fixing with checking if JIT is in foreground.
Also change to not call dispose in codeGen/gathering data, to prevent unpredicted reentrant issue.

Fix the real possible points of dead lock on etw rundown and remove IsLockedByAnyThread (which is hacky and not accurate, should not allow the etw rundown thread to updated the counters on function body, if any we should fix it)

### CODE REVIEW

File: lib/Runtime/Base/FunctionBody.cpp [hunk 1]
```diff
@@ -130,7 +133,7 @@ namespace Js
         }
 
         // when setting ptr to null we never need to promote
-        AutoCriticalSection aucoCS(&GlobalLock);
+        AutoCriticalSection aucoCS(GetLock());
```
Comment (dilijev) [inline lib/Runtime/Base/FunctionBody.cpp]: Was this name supposed to be `autoCS`?

File: lib/Runtime/Base/EtwTrace.cpp [hunk 2]
```diff
@@ -95,6 +95,9 @@ void EtwTrace::PerformRundown(bool start)
 
     while(threadContext != nullptr)
     {
+        // Take the auxPtrs updating lock, in case auxPtrs is expanding causes GC
+        // which locks Etw Rundown lock, hence dead lock.
+        AutoCriticalSection autoCs(FunctionProxy::GetLock());
```
Comment (dilijev) [inline lib/Runtime/Base/EtwTrace.cpp]: nit: consistency with casing of `autoCS` used elsewhere.

**Comment by dilijev** [comment]
Oh whoops, late to the party. Feel free to ignore my nits.

File: lib/Runtime/Base/FunctionBody.cpp (hunk 1 above)
Comment (leirocks) [inline lib/Runtime/Base/FunctionBody.cpp]: let me take it in my next PR. thanks!

**Pull request merged**